File: 20131209/AM/131209_tests/swallow/solutions/brute.cpp
#include <iostream>
#include <sstream>
#include <string>

using namespace std;

const int MAX_SIZE = 500;

string toString(int v) {
  ostringstream oss;
  oss << v;

  return oss.str();
}

string fold(const string& s) {
  if (s.size() == 1)
    return s;

  string ans = s;
  for (int i = 1; i < s.size(); ++i) {
    const string p1 = s.substr(0, i);
    const string p2 = s.substr(i);
    const string p1a = fold(p1);
    const string p2a = fold(p2);

    string t = p1a + p2a;
    if (t.size() < ans.size())
      ans = t;

    if (p2.size() % p1.size() == 0) {
      int idx = 0;
      bool is_repeated = true;

      while (idx != p2.size()) {
        if (p2.substr(idx, p1.size()) != p1) {
          is_repeated = false;
          break;
        }
        idx += p1.size();
      }

      if (is_repeated) {
        t = toString(p2.size()/p1.size()+1) + "(" + p1a + ")";
        if (t.size() < ans.size()) {
          ans = t;
        }
      }
    }
  }

  return ans;
}

int main(int argc, char** argv) {
  int N;
  cin >> N;
  string s;
  cin >> s;
  cout << fold(s) << endl;

  return 0;
}

File: 20131209/AM/131209_tests/swallow/solutions/brute.java
import java.io.*;
import java.util.*;

public class swallow {
  public final int MAX_SIZE = 500;


  public static String fold(String s) {
    if (s.length() == 1)
      return s;

    String ans = s;
    for (int i = 1; i < s.length(); ++i) {
      final String p1 = s.substring(0, i);
      final String p2 = s.substring(i);
      final String p1a = fold(p1);
      final String p2a = fold(p2);

      String t = p1a + p2a;
      if (t.length() < ans.length())
        ans = t;

      if (p2.length() % p1.length() == 0) {
        int idx = 0;
        boolean is_repeated = true;

        while (idx != p2.length()) {
          if (!p2.substring(idx, idx+p1.length()).equals(p1)) {
            is_repeated = false;
            break;
          }
          idx += p1.length();
        }

        if (is_repeated) {
          t = (p2.length()/p1.length()+1) + "(" + p1a + ")";
          if (t.length() < ans.length()) {
            ans = t;
          }
        }
      }
    }

    return ans;
  }

  public static void main(String args[]) throws Exception {
    BufferedReader fin = new BufferedReader(new InputStreamReader(System.in));
    int N = Integer.parseInt(fin.readLine());
    String s = fin.readLine().trim();
    System.out.println(fold(s));
  }
}
